autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py

```diff
@@ -0,0 +1,234 @@
+
+        Args:
+            data (np.ndarray): input features to the network
+            targets (np.ndarray): ground truth to calculate loss
+
+        Returns:
+            adv_data (np.ndarray): the adversarial examples.
+        """
+        data_copy = deepcopy(data)
+        data_copy = data_copy.float().to(self.device)
+        targets = targets.long().to(self.device)
+        data_copy = torch.autograd.Variable(data_copy)
+        data_copy.requires_grad = True
+
+        outputs = self.model(data_copy)
+        cost = self.criterion(outputs, targets)
+
+        grad = torch.autograd.grad(cost, data_copy, retain_graph=False, create_graph=False)[0]
+
+        adv_data = data_copy + self.epsilon * grad.sign()
+        adv_data = torch.clamp(adv_data, min=0, max=1).detach()
+
+        return adv_data
+
+    @staticmethod
+    def get_properties(dataset_properties: Optional[Dict[str, Any]] = None
+                       ) -> Dict[str, Union[str, bool]]:
+
+        return {
+            'shortname': 'AdversarialTrainer',
+            'name': 'AdversarialTrainer',
+            'handles_tabular': True,
+            'handles_image': False,
+            'handles_time_series': False,
+        }
+
+    @staticmethod
+    def get_hyperparameter_search_space(
+        dataset_properties: Optional[Dict] = None,
+        weighted_loss: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="weighted_loss",
+            value_range=[True, False],
+            default_value=True),
+        la_steps: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="la_steps",
+            value_range=(5, 10),
+            default_value=6,
+            log=False),
+        la_alpha: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="la_alpha",
+            value_range=(0.5, 0.8),
+            default_value=0.6,
+            log=False),
+        use_lookahead_optimizer: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="use_lookahead_optimizer",
+            value_range=(True, False),
+            default_value=True),
+        use_stochastic_weight_averaging: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="use_stochastic_weight_averaging",
+            value_range=(True, False),
+            default_value=True),
+        use_snapshot_ensemble: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="use_snapshot_ensemble",
+            value_range=(True, False),
+            default_value=True),
+        se_lastk: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="se_lastk",
+            value_range=(3,),
+            default_value=3),
+        epsilon: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter="epsilon",
+            value_range=(0.05, 0.2),
+            default_value=0.2),
```

Comment:
I do not know the good range for this parameter, but shouldn't the range be larger and we should apply a log-scale for this param.